core/src/main/java/hudson/FilePath.java

```diff
@@ -3264,7 +3264,13 @@ private File deleting(File f) {
         return f;
     }
 
-    private boolean mkdirs(File dir) throws IOException {
+    @Restricted(NoExternalUse.class)
+    boolean mkdirs(File dir) throws IOException {
+        // when checking for existence, we are in fact following symlinks, so we need to be consistent
+        // and create the dir, for the link we followed
+        if(Files.isSymbolicLink(dir.toPath())) {
+            dir = Files.readSymbolicLink(dir.toPath()).toFile();
```

Comment:
Use `Util#fileToPath` instead of `Path#toPath` as mentioned above. Maybe better to use `Util#isSymlink` instead of `Files#isSymbolicLink` to pick up Windows junction points (`mklink /J`) correctly, although I don't think those can be resolved like other symlinks so it might get messy.

Regardless of those points, this change seems a little concerning to me. My first thought would be to instead change `dir.exists()` to `Files.exists(fileToPath(dir), LinkOption.NOFOLLOW_LINKS)` so that we don't follow symlinks in the first place, although we'd need to run the test suite to see if that is feasible.

Reply:
I'll wait on an answer on the test question before changing anything in this method.